src/Microsoft.Health.Fhir.SqlServer/Features/Watchdogs/DefragWatchdog.cs

```diff
@@ -7,193 +7,302 @@
 using System.Data;
 using System.Linq;
-using System.Security.Cryptography;
 using System.Threading;
 using System.Threading.Tasks;
+using EnsureThat;
+using MediatR;
 using Microsoft.Data.SqlClient;
 using Microsoft.Extensions.Logging;
 using Microsoft.Health.Extensions.DependencyInjection;
-using Microsoft.Health.Fhir.Core.Features.Operations;
+using Microsoft.Health.Fhir.Core.Messages.Storage;
 using Microsoft.Health.Fhir.SqlServer.Features.Schema;
 using Microsoft.Health.Fhir.SqlServer.Features.Schema.Model;
 using Microsoft.Health.Fhir.SqlServer.Features.Storage;
 using Microsoft.Health.JobManagement;
 using Microsoft.Health.SqlServer.Features.Client;
 using Microsoft.Health.SqlServer.Features.Schema;
+using Polly;
 
 namespace Microsoft.Health.Fhir.SqlServer.Features.Watchdogs
 {
-    public class DefragWatchdog : WatchdogTimer
+    public sealed class DefragWatchdog : INotificationHandler<StorageInitializedNotification>
     {
-        private const byte _queueType = (byte)QueueType.Defrag;
+        private const byte QueueType = (byte)Core.Features.Operations.QueueType.Defrag;
+        private const string PeriodHourId = "Defrag.Period.Hours";
+        internal const string IsEnabledId = "Defrag.IsEnabled";
+        private const string HeartbeatTimeoutSecId = "Defrag.HeartbeatTimeoutSec";
+        private const string HeartbeatPeriodSecId = "Defrag.HeartbeatPeriodSec";
+        private const string ThreadsId = "Defrag.Threads";
+
         private int _threads;
-        public const string ThreadsId = "Defrag.Threads";
         private int _heartbeatPeriodSec;
-        public const string HeartbeatPeriodSecId = "Defrag.HeartbeatPeriodSec";
         private int _heartbeatTimeoutSec;
-        public const string HeartbeatTimeoutSecId = "Defrag.HeartbeatTimeoutSec";
-        private double _periodSec;
-        public const string PeriodSecId = "Defrag.PeriodSec";
-        public const string IsEnabledId = "Defrag.IsEnabled";
-        private SqlConnectionWrapperFactory _sqlConnectionWrapperFactory;
-        private SchemaInformation _schemaInformation;
-        private SqlQueueClient _sqlQueueClient;
-        private ILogger<DefragWatchdog> _logger;
-
-        public DefragWatchdog(SchemaInformation schemaInformation, Func<IScoped<SqlConnectionWrapperFactory>> sqlConnectionWrapperFactory, Func<IScoped<SqlQueueClient>> sqlQueueClient, ILogger<DefragWatchdog> logger)
-        {
-            _schemaInformation = schemaInformation;
-            _sqlConnectionWrapperFactory = sqlConnectionWrapperFactory.Invoke().Value;
-            _sqlQueueClient = sqlQueueClient.Invoke().Value;
-            _logger = logger;
-        }
+        private double _periodHour;
 
-        public void Start()
+        private readonly Func<IScoped<SqlConnectionWrapperFactory>> _sqlConnectionWrapperFactory;
+        private readonly SchemaInformation _schemaInformation;
+        private readonly Func<IScoped<SqlQueueClient>> _sqlQueueClient;
```

Comment:
What is the reason for applying .Invoke().Value logic outside of constructor?

Reply:
`.Invoke().Value` would mean the resolved instance will live as along as the owning class, since the owning class is Singleton this lifetime would be dangerous for scoped components.

Reply:
Could you explain how is this different from having defrag watchdog instance living forever?